### ISSUE TITLE
GPU memory still allocated when fastNlMeansDenoising() returns

### ISSUE THREAD

**Issue by angelo-peronio** [opencv/opencv#17789]
##### System information (version)

- OpenCV => ~~4.1.1~~ 4.3 (installed using `vcpkg`)
- Operating System / Platform => Windows 10 64 Bit
- Compiler => Visual Studio 2019 version 16.6.5
- GPU => NVIDIA GeForce GTX 1050 with 4 GiB dedicated video memory
- NVIDIA driver version => 446.14

##### Detailed description

I am running `cv::fastNlMeansDenoising()` on the GPU via T-API / `UMat`s / OpenCL. I am not able to loop over a series of images of the same size: after a few iterations the calculation falls back on the CPU with the error message `CL_MEM_OBJECT_ALLOCATION_FAILURE (-4)`. It seems that when `cv::fastNlMeansDenoising()` returns, some GPU memory has not been deallocated yet.

##### Steps to reproduce

Consider the following code:

```.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <chrono>
#include <thread>

using namespace std::chrono_literals;
using namespace std::string_literals;

void loopNlMeans()
{
    auto const imageWidth = 3500;
    auto const h = 14.0;
    auto const kernelSizeParam = 15;
    auto const nLoops = 8;

    auto image = cv::UMat::zeros(imageWidth, imageWidth, CV_8UC1);
    auto dst = cv::UMat::zeros(imageWidth, imageWidth, CV_8UC1);
    cv::theRNG().state = 7;
    cv::randu(image, 0, 256);

    for (int i = 0; i < nLoops; ++i)
    {
        std::cout << "Loop "s << i << "\n"s;
        cv::fastNlMeansDenoising(image, dst, h, kernelSizeParam, kernelSizeParam);
    }
}

int main()
{
    loopNlMeans();
}
```

I would expect it to run completely on the GPU, whereas I get the following output:

```txt
Loop 0
Loop 1
Loop 2
Loop 3
OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueNDRangeKernel('fastNlMeansDenoising', dims=2, globalsize=28160x110x1, localsize=256x1x1) sync=false
```

If I add `std::this_thread::sleep_for(1s);` inside the `for` loop, the calculation goes on for more iterations before falling back to the CPU:

```txt
Loop 0
Loop 1
Loop 2
Loop 3
Loop 4
Loop 5
Loop 6
OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueNDRangeKernel('fastNlMeansDenoising', dims=2, globalsize=28160x110x1, localsize=256x1x1) sync=false
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [ ] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by angelo-peronio** [comment]
It seems that declaring and copying to a small `UMat` in the loop is a viable workaround:

```cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace std::string_literals;

void loopNlMeans()
{
    auto const imageWidth = 3500;
    auto const h = 14.0;
    auto const kernelSizeParam = 15;
    auto const nLoops = 15;

    auto image = cv::UMat::zeros(imageWidth, imageWidth, CV_8UC1);
    auto dst = cv::UMat::zeros(imageWidth, imageWidth, CV_8UC1);
    cv::theRNG().state = 7;
    cv::randu(image, 0, 256);

    for (int i = 0; i < nLoops; ++i)
    {
        std::cout << "Loop "s << i << "\n"s;
        cv::fastNlMeansDenoising(image, dst, h, kernelSizeParam, kernelSizeParam);
        cv::UMat trick;
        cv::UMat::zeros(1, 1, CV_8UC1).copyTo(trick);
    }
}

int main()
{
    loopNlMeans();
}
```

No more OpenCL allocation failures:
```text
Loop 0
Loop 1
Loop 2
Loop 3
Loop 4
Loop 5
Loop 6
Loop 7
Loop 8
Loop 9
Loop 10
Loop 11
Loop 12
Loop 13
Loop 14

C:\...\opencv-opencl-tests\x64\Debug\opencv-opencl-tests.exe (process 20252) exited with code 0.
Press any key to close this window . . .
```

🤔

**Comment by alalek** [comment]
OpenCL / T-API is asynchronous API by design.
Most part of calls are usually just allocate / schedule kernels. No real processing done at this point (performed in background).

So, your first example schedules multiple tasks at once (expecting processing in parallel) which high pressure on available resources.

OpenCL synchronization points are:
- fetching data on CPU through `UMat::getMat()`. This is expected way how application should work.
- or forced waiting for OpenCL queue using `cv::ocl::finish()` (`#include <opencv2/core/ocl.hpp>`).

---

Usage questions should go to Users OpenCV Q/A forum: http://answers.opencv.org

**Comment by angelo-peronio** [comment]
Thank you! Everything makes much more sense now.

Sorry for asking in the wrong place, I was not understanding the symptoms I saw, and suspected a bug.

My use case: process a big `UMat` by tiling it in pieces small enough to be processed with the available GPU memory.

**Comment by angelo-peronio** [comment]
`cv::ocl::finish()` does indeed wait for the GPU to finish, but it seems that it does not wait for memory deallocation.

More details on [Stackoverflow](https://stackoverflow.com/q/63110817).

**Comment by mshabunin** [comment]
I may be wrong, but buffers will not be removed automatically, because they are kept in a buffer pool. Try to [get OCL allocator](https://docs.opencv.org/master/dc/d83/group__core__opencl.html#ga2de08873daa452bb88f17d5054aa3aae), then [BufferPoolController](https://docs.opencv.org/master/df/d4c/classcv_1_1MatAllocator.html#a22e35b7a843280c15308133f6787fdb6), then either reduce its size, either free all buffers (see [documentation](https://docs.opencv.org/master/d4/dad/classcv_1_1BufferPoolController.html)).

**Comment by angelo-peronio** [comment]
Thank you for the idea! I think I am missing something obvious here. I implemented your suggestions, adapting my Stackoverflow [question](https://stackoverflow.com/q/63110817):

```c++
#include <opencv2/opencv.hpp>
#include <opencv2/core/ocl.hpp>
#include <iostream>

int main()
{
    auto const nLoops = 5;
    auto const imageWidth = 46340;  // Image size ~ 2 GiB.
    auto* const allocatorPtr = cv::ocl::getOpenCLAllocator();
    auto* const bpcPtr = allocatorPtr->getBufferPoolController();
    std::cout << "\n";

    for (int iLoop = 0; iLoop < nLoops; ++iLoop)
    {
        std::cout << "Loop " << iLoop << " begins.\n";
        {
            // bigImage will be destroyed as soon as it is out of scope.
            auto const bigImage = cv::UMat::zeros(imageWidth, imageWidth, CV_8UC1);
        }
        std::cout << "Buffer pool controller reserved size: " << bpcPtr->getReservedSize() << " B.\n";
        std::cout << "Buffer pool controller maximum reserved size: " << bpcPtr->getMaxReservedSize() << " B.\n";
        cv::ocl::finish();
        bpcPtr->freeAllReservedBuffers();
        std::cout << "\n";
    }

    std::cout << "Success!";
}
```

I get the following — notice that I have updated to OpenCV 4.3:

```text
[ INFO:0] global C:\tools\vcpkg\buildtrees\opencv4\src\4.3.0-0c6047baf6.clean\modules\core\src\ocl.cpp (891) cv::ocl::haveOpenCL Initialize OpenCL runtime...

Loop 0 begins.
[ INFO:0] global C:\tools\vcpkg\buildtrees\opencv4\src\4.3.0-0c6047baf6.clean\modules\core\src\ocl.cpp (433) cv::ocl::OpenCLBinaryCacheConfigurator::OpenCLBinaryCacheConfigurator Successfully initialized OpenCL cache directory: C:\Users\ANGELO~1.PER\AppData\Local\Temp\opencv\4.3\opencl_cache\
[ INFO:0] global C:\tools\vcpkg\buildtrees\opencv4\src\4.3.0-0c6047baf6.clean\modules\core\src\ocl.cpp (457) cv::ocl::OpenCLBinaryCacheConfigurator::prepareCacheDirectoryForContext Preparing OpenCL cache configuration for context: NVIDIA_Corporation--GeForce_GTX_1050--446_14
Buffer pool controller reserved size: 0 B.
Buffer pool controller maximum reserved size: 0 B.

Loop 1 begins.
OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueNDRangeKernel('set', dims=2, globalsize=11776x46344x1, localsize=NULL) sync=false
OpenCV(4.3.0) Error: Unknown error code -220 (OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueReadBuffer(q, handle=000002041E63D650, CL_TRUE, 0, sz=2147395600, data=00000204956CC080, 0, 0, 0)) in cv::ocl::OpenCLAllocator::map, file C:\tools\vcpkg\buildtrees\opencv4\src\4.3.0-0c6047baf6.clean\modules\core\src\ocl.cpp, line 5138
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.3.0) Error: Unknown error code -220 (OpenCL error CL_MEM_OBJECT_ALLOCATION_FAILURE (-4) during call: clEnqueueReadBuffer(q, handle=000002041E63D650, CL_TRUE, 0, sz=2147395600, data=00000204956CC080, 0, 0, 0)) in cv::ocl::OpenCLAllocator::map, file C:\tools\vcpkg\buildtrees\opencv4\src\4.3.0-0c6047baf6.clean\modules\core\src\ocl.cpp, line 5138
```

* If the GPU memory is kept in a buffer pool after the destruction of `bigImage` during loop 0, why is it not available to allocate `bigImage` during loop 1? It could be that I am misunderstanding the concept of buffer pool, however.
* The maximum reserved size of the buffer pool seems to be 0 B. This surprised me, and prevents me to reduce its size.
* Freeing all the reserved buffers does not help — probably also because they are 0 B big.

**Comment by alalek** [comment]
> cv::ocl::finish() does indeed wait for the GPU to finish, but it seems that it does not wait for memory deallocation.

Cleanup of OpenCL buffers is performed (scheduled) in OpenCL kernel's completion callbacks.
`clFinish()` has no requirement to call the "ready" callbacks.
This can't be fixed in OpenCV.

**Comment by angelo-peronio** [comment]
@alalek Thank you. I understand from your [comment](https://github.com/opencv/opencv/issues/17789#issuecomment-743979206) that `cv::ocl::finis()` is *not* there to ensure that GPU memory is deallocated after a T-API call.

My question was: is there any way to do that?

I understand you would rather have such usage questions on the OpenCV forum, so I asked [there](https://forum.opencv.org/t/338) too.
